src/Utilities/FileSystem.cpp

```diff
@@ -0,0 +1,291 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include "Utilities/FileSystem.hpp"
+
+#include <cerrno>
+#include <chrono>
+#include <cmath>
+#include <cstdlib>
+#include <dirent.h>
+#include <libgen.h>
+#include <regex>
+#include <sstream>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <thread>
+#include <unistd.h>
+
+#include "ErrorHandling/Error.hpp"
+#include "Parallel/Printf.hpp"
+
+namespace filesystem {
+
+std::string dirname(const std::string& path) {
+  auto* file_path = new char[path.length() + 1];
+  strncpy(file_path, path.c_str(), path.length() + 1);
+  // The pointer from ::dirname is not deleted since it aliases with file_path
+  char* parent_dir_name = ::dirname(file_path);
+  std::string return_name(parent_dir_name);
+  delete[] file_path;
+  return return_name;
+}
+
+std::string get_file_name(const std::string& file_path) {
+  if (file_path.empty()) {
+    ERROR("Received an empty path");
+  }
+  if (file_path.find('/') == std::string::npos) {
+    // Handle file names such as 'dummy.txt' or '.dummy.txt'
+    return file_path;
+  }
+  std::smatch match{};
+  std::regex file_name_pattern{R"(^.*/(.+\..+))"};
+  auto regex_matched = std::regex_search(file_path, match, file_name_pattern);
+  if (not regex_matched) {
+    ERROR("Failed to find a file in the given path: '" << file_path << "'");
+  }
+  return match[1];
+}
+
+std::string get_absolute_path(const std::string& rel_path) {
+  char* name = realpath(rel_path.c_str(), nullptr);
+  if (nullptr == name) {
+    if (ENAMETOOLONG == errno) {
+      // LCOV_EXCL_START
+      ERROR(
+          "Failed to convert to absolute path because the resulting name is "
+          "too long. Relative path is: '"
+          << rel_path << "'.");
+      // LCOV_EXCL_STOP
+    } else if (EACCES == errno) {
+      // LCOV_EXCL_START
+      ERROR(
+          "Failed to convert to absolute path because one of the components of "
+          "the path does not have proper read access. Relative path is: '"
+          << rel_path << "'.");
+      // LCOV_EXCL_STOP
+    } else if (ENOENT == errno) {
+      ERROR(
+          "Failed to convert to absolute path because one of the path "
+          "components does not exist. Relative path is: '"
+          << rel_path << "'.");
+      // LCOV_EXCL_START
+    } else if (ELOOP == errno) {
+      ERROR(
+          "Failed to convert to absolute path because the maximum number of "
+          "symlinks was in the path. Relative path is: '"
+          << rel_path << "'.");
+      // LCOV_EXCL_STOP
+    }
+    ERROR("Failed to get absolute path for an unknown reason.");
```

Comment:
Include `error` in the message.  Probably has to be saved since the `ERROR` internals might overwrite it.